session.go

```diff
@@ -973,24 +973,69 @@ func (st *StmtTxn) SeekReverse(k kv.Key) (kv.Iterator, error) {
 	return kv.NewUnionIter(bufferIt, retrieverIt, true)
 }
 
-// StmtCommit implements the Transaction interface.
-func (st *StmtTxn) StmtCommit() error {
-	err := kv.WalkMemBuffer(st.buf, func(k kv.Key, v []byte) error {
+// StmtCommit implements the context.Context interface.
+func (s *session) StmtCommit() error {
+	st := &s.StmtTxn
+	err := kv.WalkMemBuffer(s.buf, func(k kv.Key, v []byte) error {
 		if len(v) == 0 {
 			return errors.Trace(st.Transaction.Delete(k))
 		}
 		return errors.Trace(st.Transaction.Set(k, v))
 	})
 	st.buf.Reset()
+
+	// Need to flush binlog.
```

Comment:
I think we should check error before merge binlog.

Reply:
check what error? StmtCommit means there is no error.

Reply:
!!! Delete or Set error.

Reply:
➜  tidb git:(statement-binlog) ✗ cat ttt
1. `StmtCommit` be called only if no error happen during executing a statement.
2. Whether `StmtCommit` return error or not, it should reset related cache.
3. `StmtCommit` should also makes no side effect when it meets an error.

Point 2 is important, and **Point 3 is hard**, if you take a close look on it.
@jackysp @winkyao

Reply:
got